proposals/list-patterns.md

```diff
@@ -0,0 +1,122 @@
+
+slice_pattern
+  : '..' negated_pattern?
+	
+primary_pattern
+  : recursive_pattern
+  | slice_pattern
+
+```
+There are two new additions to the *recursive pattern* syntax as well as a *slice pattern*: 
+
+- The *list pattern clause* is used to match elements and the *length pattern clause* is used to match the length.
+- A _slice pattern_ is only permitted once and only directly in a *list pattern* and discards _**zero or more**_ elements.
+
+
+Due to the ambiguity with *property pattern clause*, the *list pattern clause* cannot be empty and a *length pattern* should be used. 
+
+
+
+
+
+
+#### Pattern compatibility
+
+A *length pattern* is compatible with any type that conforms to the following rules:
+
+1. Has an accessible property getter that returns an `int` and has the name `Length` or `Count`
+
+A *list pattern* is compatible with any type that conforms to the following rules:
+
+1. Is compatible with the length pattern
+2. Has an accessible indexer with a single `int` parameter
+
+A *slice pattern* is compatible with any type that conforms to the following rules:
+
+1. Is compatible with the list pattern
+2. Has an accessible `Slice` method that takes two `int` parameters (required only if a subpattern is specified)
+
+This set of rules is already specified as the ***range indexer pattern*** and required for range indexers.
+
+#### Subsumption checking
+
+  Subsumption checking works just like recursive patterns with `ITuple` - corresponding subpatterns are matched by position plus an additional node for testing length.
+
+#### Lowering
+
+ A pattern of the form `expr is {1, 2, 3}` is equivalent to the following code:
+```cs
+expr.Length is 3
+&& expr[0] is 1
+&& expr[1] is 2
+&& expr[2] is 3
+```
+A *slice pattern* acts like a proper discard i.e. no tests will be emitted for such pattern, rather it only affects other nodes, namely the length and indexer. For instance, a pattern of the form `expr is {1, .. var s, 3}`  is equivalent to the following code: 
+```cs
+expr.Length    is >= 2
+&& expr[0]     is 1
+&& expr[1..^1] is var s
+&& expr[^1]    is 3
+```
+Note: the lowering is presented in the pattern form here to show how subsumption checking works, for example, the following code produces an error because both patterns yield the same DAG:
+
+```cs
+case {_, .., 1}: // expr.Length is >= 2 && expr[^1] is 1
+case {.., _, 1}: // expr.Length is >= 2 && expr[^1] is 1
+```
+Unlike:
+```cs
+case {_, 1, ..}: // expr.Length is >= 2 && expr[1] is 1
+case {.., 1, _}: // expr.Length is >= 2 && expr[^2] is 1
+```
+
+Note: The pattern `{..}` lowers to `expr.Length >= 0` so it would not be considered as a catch-all.
+
+### Additional types
+
+Beyond the pattern-based mechanism outlined above, there are an additional two set of types that can be covered as a special case.
+
+- **Multi-dimensional arrays**: All nested list patterns must agree to a length range. 
+- **Foreach-able types**: This includes pattern-based and extension `GetEnumerator`.
```

Comment:
I did not go through the details on these. I can update later when I work out the specifics.